File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& arr) {
        int n = arr.size();
        int profit = 0;
        int curr_deal = 0;
        int mini = arr[0];
        for(int i = 1; i<n; i++){
            curr_deal = arr[i] - mini;
            profit = max(profit, curr_deal);
            mini = min(mini, arr[i]);
        }
        return profit;
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int curr_deal = 0;
        int profit_mx = 0;
        int min_buy = prices[0];
        for(int i = 1; i<n; i++){
            curr_deal = prices[i] - min_buy;
            profit_mx = Math.max(profit_mx, curr_deal);
            min_buy = Math.min(min_buy, prices[i]);
        }
        return profit_mx;
    }
}